backends/dnf/pk-backend-dnf.c

```diff
@@ -94,24 +97,37 @@ pk_backend_supports_parallelization (PkBackend *backend)
 	return FALSE;
 }
 
+static gboolean
+pk_backend_check_sack_timer (gpointer key, gpointer value, gpointer user_data)
+{
+	DnfSackCacheItem *cache_item = value;
+	if (g_timer_elapsed (cache_item->timer, NULL) > SACK_TTL) {
+		g_debug ("invalidating %s as expired", (char *)key);
+		return TRUE;
+	}
+	return FALSE;
+}
+
+static gboolean
+pk_backend_sack_expire (gpointer user_data)
+{
+	PkBackendDnfPrivate *priv = user_data;
+	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->sack_mutex);
+
+	g_hash_table_foreach_remove (priv->sack_cache,
+				     pk_backend_check_sack_timer, NULL);
+	return TRUE;
+}
+
 static void
 pk_backend_sack_cache_invalidate (PkBackend *backend, const gchar *why)
 {
-	GList *l;
-	DnfSackCacheItem *cache_item;
 	PkBackendDnfPrivate *priv = pk_backend_get_user_data (backend);
-	g_autoptr(GList) values = NULL;
 	g_autoptr(GMutexLocker) locker = g_mutex_locker_new (&priv->sack_mutex);
```

Comment:
does this suggest that it's not always single threaded? I can't remember the details.

Reply:
It's not single theaded, but removing a DnfSackCacheItem from sack_cache should be safe, since it only unrefs the dnf sack that the CacheItem references.  The thread that requested the dnf sack will unref and free the sack when it is done with it.